drivers: flash: Fix flash_write_protection_set usage

eugmes opened the pull request:
This PR cleans up usage of `flash_write_protection_set()`. The second argument of this function as `bool` so passing 0 and 1 is not entirely correct. The PR also changes `flash_shell.c` so that flash protection is disabled before each flash erase or write operation because drivers may automatically enable flash protection once such operation is complete.

eugmes committed:
drivers: flash: Pass bool to flash_write_protection_set()

The second argument of this foonction is a bool, so passing 0 and 1
is incorrect.

Coccinelle script:

    @@
    expression e;
    @@
    flash_write_protection_set(e,
    (
    - 0
    + false
    |
    - 1
    + true
    )
     )

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

eugmes committed:
drivers: flash_shell: Fix write protection in the test command

Flash drivers are free to re-enable write protection after a write or
erase operation is complete. Therefore write protection has to be
disabled before any such operation.

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

Laczen reviewed (approved):
Thanks for the improvement.

ioannisg commented on drivers/flash/flash_shell.c:
```diff
@@ -193,13 +193,13 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
 	}
 
 	while (repeat--) {
+		flash_write_protection_set(flash_dev, false);
```
Just wondering: shouldn't write protection be "restored" before (or right after) exiting the while loop?

@nvlsianpu ^^

nvlsianpu replied:
In that case should be restored after while loop.

In general recommendation is to set flash protecting after packet of write/erase operation was completed.

nvlsianpu replied:
@eugmes ^^

eugmes replied:
@nvlsianpu I can fix this, but the original code also didn't restore write protection.

eugmes replied:
Fixed. It's either `goto` or duplicate code.

nvlsianpu replied:
thanks!

ioannisg replied:
wonder if we really need to have "goto" here, isn't there a refactor that removes the need, without duplicating code too much?

eugmes replied:
@ioannisg How about this?

eugmes replied:
oops.

eugmes replied:
Now.

eugmes committed:
drivers: flash: Restore write protection in flash_shell

Restore the flash write proectction after disabling it for
flash write and erase shell commands.

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

ioannisg reviewed (approved):
LGTM
@nvlsianpu , maybe another quick look?
